quickwit/quickwit-search/src/fetch_docs.rs

```diff
@@ -240,8 +242,37 @@ async fn fetch_docs_in_split(
         }
     });
 
-    let stream = futures::stream::iter(doc_futures).buffer_unordered(NUM_CONCURRENT_REQUESTS);
-    stream.try_collect::<Vec<_>>().await
+    spawn_buffer_unordered(doc_futures, NUM_CONCURRENT_REQUESTS).await
+}
+
+/// Run a group a future in a buffered maner, spawning each one in a dedicated task.
+///
+/// Stop processing if any future returns an `Err`.
+async fn spawn_buffer_unordered<T, E, Fut>(
+    mut futures: impl Iterator<Item = Fut>,
+    max_running: usize,
+) -> Result<Vec<T>, E>
+where
+    Fut: Future<Output = Result<T, E>> + Send + 'static,
+    T: Send + 'static,
+    E: Send + 'static,
+{
+    assert!(max_running > 0);
+
+    let mut results = Vec::new();
+    let mut future_set = futures::stream::FuturesUnordered::new();
+    for future in (&mut futures).take(max_running) {
+        future_set.push(tokio::spawn(future.in_current_span()));
+    }
+
+    while let Some(result) = future_set.next().await {
```

Comment:
You might as well use 
https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffer_unordered

directly I think...

Reply:
hum, actually you are right, iter are lazy so `map()` won't call `tokio::spawn` until `buffer_unordered` tries to schedule more tasks